src/abstractdataframe/join.jl

```diff
@@ -91,6 +91,291 @@ _rename_cols(old_names::AbstractVector{Symbol},
            (renamecols isa Function ? Symbol(renamecols(string(n))) : Symbol(n, renamecols))
            for n in old_names]
 
+prepare_on_col() = throw(ArgumentError("at least one on column required when joining"))
+prepare_on_col(c::AbstractVector) = c
+prepare_on_col(cs::AbstractVector...) = tuple.(cs...)
+
+# in map2refs zero(V) is PooledArray.jl specific
+function map2refs(x::AbstractVector, ref::PooledArray{T,V,1}) where {T, V}
+    refip = ref.invpool
+    return [get(refip, v, zero(V)) for v in x]
+end
+
+function map2refs(x::PooledVector, ref::PooledArray{T,V,1}) where {T, V}
+    refip = ref.invpool
+    mapping = [get(refip, v, zero(V)) for v in x.pool]
+    return @inbounds [mapping[r] for r in x.refs]
+end
+
+function compose_inner_table(joiner::DataFrameJoiner,
+                             makeunique::Bool,
+                             left_rename::Union{Function, AbstractString, Symbol},
+                             right_rename::Union{Function, AbstractString, Symbol})
+
+    right_shorter = length(joiner.dfr_on[!, 1]) <= length(joiner.dfl_on[!, 1])
+
+    left_cols = collect(eachcol(joiner.dfl_on))
+    right_cols = collect(eachcol(joiner.dfr_on))
+
+    disallow_sorted = false
+
+    if right_shorter
+        for i in eachindex(left_cols, right_cols)
+            rc = right_cols[i]
+            lc = left_cols[i]
+            if lc isa PooledArray
+                right_cols[i] = map2refs(rc, lc)
+                left_cols[i] = lc.refs
+            end
+        end
```

Comment:
This loop is a no-op when `lc` isn't a `PooledArray`, right?

Reply:
yes, but this is a small loop (it loops over columns we join-on). I will replace this code anyway when `DataAPI.invrefpool` is supported.

Reply:
How about putting the `if lc isa PooledArray` around the `for`?

Reply:
but `lc` is defined inside the `for` loop.

Essentially what we do here is:
1. iterate all pairs of columns we join on; each pair is `lc` and `rc` column
2. if left - longer - colunm (`lc`) is `PooledArray` (in the future `DataAPI.refarray` compliant) then we replace it by refs, and we appropriately map the shorter column `rc` to refs from longer column (as we want to save time, so we map the shorter column not the longer one)